### PR TITLE
CORE: fix score update when only score given

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Fixes ucc score table when user provides score for TL/CL and only total score is given
e.g. UCC_TL_SHARP_TUNE=inf results in scatter, allgather, alltoall collectives (not supported by sharp) in UCC score table

<details>
  <summary>with fix</summary>

```bash
       ucc_team.c:471  UCC  INFO  ===== COLL_SCORE_MAP (team_id 32768, size 2) =====
ucc_coll_score_map.c:201  UCC  INFO  Allgather:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..4095}:TL_UCP:10 {4K..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Allgatherv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Allreduce:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..4095}:TL_SHARP:10 {4K..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Alltoall:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..257}:TL_UCP:10 {258..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Alltoallv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Barrier:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Bcast:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Fanin:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Fanout:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Gather:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Gatherv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce_scatter:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce_scatterv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
ucc_coll_score_map.c:201  UCC  INFO  Scatterv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_UCP:10
       ucc_team.c:474  UCC  INFO  ================================================
```
</details>

<details>
  <summary>without fix</summary>

```bash
       ucc_team.c:472  UCC  INFO  ===== COLL_SCORE_MAP (team_id 32771, size 2) =====
ucc_coll_score_map.c:201  UCC  INFO  Allgather:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..4095}:TL_SHARP:10 {4K..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Allgatherv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Allreduce:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..4095}:TL_SHARP:10 {4K..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Alltoall:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..257}:TL_SHARP:10 {258..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Alltoallv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Barrier:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Bcast:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Fanin:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Fanout:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Gather:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Gatherv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce_scatter:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Reduce_scatterv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Scatter:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO  Scatterv:
ucc_coll_score_map.c:201  UCC  INFO        Host: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Cuda: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        CudaManaged: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        Rocm: {0..inf}:TL_SHARP:10
ucc_coll_score_map.c:201  UCC  INFO        RocmManaged: {0..inf}:TL_SHARP:10
       ucc_team.c:474  UCC  INFO  ================================================
```
</details>

## How ?
Pass TL supported collectives to tuning string parser

### CODE REVIEW

File: src/components/cl/hier/cl_hier_team.c [hunk 1]
```diff
@@ -353,10 +362,14 @@ ucc_status_t ucc_cl_hier_team_get_scores(ucc_base_team_t   *cl_team,
     }
 
     for (i = 0; i < UCC_CL_HIER_N_DEFAULT_ALG_SELECT_STR; i++) {
+        // status = ucc_coll_score_update_from_str(
+        //     ucc_cl_hier_default_alg_select_str[i], score,
```
Comment (vspetrov) [inline src/components/cl/hier/cl_hier_team.c:366]: remove commented code

File: src/components/cl/hier/cl_hier_team.c [hunk 2]
```diff
@@ -365,10 +378,12 @@ ucc_status_t ucc_cl_hier_team_get_scores(ucc_base_team_t   *cl_team,
     }
 
     if (strlen(ctx->score_str) > 0) {
-        status = ucc_coll_score_update_from_str(
-            ctx->score_str, score, UCC_CL_TEAM_SIZE(team), NULL, cl_team,
-            UCC_CL_HIER_DEFAULT_SCORE, ucc_cl_hier_alg_id_to_init, NULL, 0);
-
+        // status = ucc_coll_score_update_from_str(
+        //     ctx->score_str, score, UCC_CL_TEAM_SIZE(team), NULL, cl_team,
+        //     UCC_CL_HIER_DEFAULT_SCORE, ucc_cl_hier_alg_id_to_init, NULL, 0,
```
Comment (vspetrov) [inline src/components/cl/hier/cl_hier_team.c:383]: remove commented code

**Review by vspetrov — APPROVED** [review]

File: src/coll_score/ucc_coll_score.h [hunk 3]
```diff
@@ -19,6 +19,30 @@
 
 #define UCC_MSG_MAX UINT64_MAX
 
+/* Callback that maps alg_id (int or str) to the "init" function.
+   This callback is provided by the component (CL/TL) that uses
+   ucc_coll_score_alloc_from_str.
+   Return values:
+   UCC_OK - input alg_id can be correctly mapped to the "init" fn
+   UCC_ERR_NOT_SUPPORTED - CL/TL does allow changing algorithms ids for
```
Comment (shimmybalsam) [inline src/coll_score/ucc_coll_score.h:27]: typo - shouod be "doesn't" instead of "does"

File: src/coll_score/ucc_coll_score.h [hunk 4]
```diff
@@ -112,15 +123,12 @@ ucc_status_t ucc_coll_score_alloc_from_str(const char *            str,
...
-                                            ucc_rank_t              team_size,
-                                            ucc_base_coll_init_fn_t init,
-                                            ucc_base_team_t        *team,
-                                            ucc_score_t             def_score,
-                                            ucc_alg_id_to_init_fn_t alg_fn,
-                                            ucc_memory_type_t      *mtypes,
-                                            int                     mt_n);
+
+ucc_status_t
+ucc_coll_score_update_from_str(const char *str,
```
Comment (shimmybalsam) [inline src/coll_score/ucc_coll_score.h:128]: Once merged, needs to be updated in TL/SHM as well.

File: test/gtest/coll_score/test_score_update.cc [hunk 5]
```diff
@@ -34,7 +34,8 @@ UCC_TEST_F(test_score_update, non_overlap)
                UCC_COLL_TYPE_BARRIER);
     init_score(update, RLIST({RANGE(10, 20, 100), RANGE(30, 35, 1)}),
                UCC_COLL_TYPE_BARRIER);
-    EXPECT_EQ(UCC_OK, ucc_coll_score_update(score, update, 0, NULL, 0));
+    EXPECT_EQ(UCC_OK, ucc_coll_score_update(score, update, 0, NULL, 0,
+              UCC_COLL_TYPE_ALL));
```
Comment (shimmybalsam) [inline test/gtest/coll_score/test_score_update.cc:38]: Will UCC_COLL_TYPE_ALL work for TL/SHM as well? Not all colls are supported there - will it fall back onto other TL?

**Review by shimmybalsam — APPROVED** [review]

**Review by manjugv — APPROVED** [review]

**Pull request merged**